[MRG+1] Adding various averages to classification_report

apacha opened the pull request:
#### Reference Issues/PRs
Fixes #4558
Closes #4622

#### What does this implement/fix? Explain your changes.
Adding average-option to classification_report to make default more explicit and allow changing it easily for the report.

#### Any other comments?
Added tests and updated existing ones. Note that the output now has slightly changed for the average line from `avg / total` to `average_method avg`, e.g. `weighted avg`.

apacha committed:
Adding average-option to classification_report to make default more explicit and allow changing it easily for the report (#4558).

apacha committed:
Fixed documentation to reflect new output.

apacha committed:
Replacing remaining occurencies of total/avg in documentation with "weighted avg" and improving style to comply with Style-guidelines (line-width <80 chars).

jnothman commented on sklearn/metrics/classification.py:
```diff
@@ -1451,14 +1451,39 @@ def classification_report(y_true, y_pred, labels=None, target_names=None,
+        reporting the individual results per class:
+
+        ``'binary'``:
+            Only report results for the class specified by ``pos_label``.
```
This isn't applicable here. We report scores for all binarizations.

jnothman commented on sklearn/metrics/classification.py:
```diff
@@ -1517,11 +1547,16 @@ class 2       1.00      0.67      0.80         3
-    # compute averages
+    # compute averages with specified averaging method
+    avg_p, avg_r, avg_f1, unused_s = \
+        precision_recall_fscore_support(y_true, y_pred, labels=labels,
```
If we're going to do it this way (despite the fact that macro-averages can be calculated easily from the above class-wise scores), why wouldn't we allow multiple values to be provided to `average`? i.e. `average=['micro', 'weighted', 'macro']`

jnothman commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -890,17 +890,106 @@ def test_multilabel_classification_report():
+         0       0.50      0.67      0.57         3
+         1       0.75      0.60      0.67         5
+
+binary avg       0.75      0.60      0.67         8
```
This is redundant. Let's not support this, please.

jnothman commented:
Thanks for the PR

apacha commented:
Thanks for the review. Will fix the remaining issues after Nov. 30th:

- [ ] Remove documentation for binary
- [ ] Remove binary option, because it is redundant
- [ ] Support multiple values to be calculated

apacha committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	sklearn/metrics/classification.py

apacha committed:
Removing 'binary' option as it is redundant.

jnothman reviewed (commented):
Sorry to say so late in the piece, but when I look at this, I think it a bit silly that we do not just report scores under all applicable averaging schemes: macro and weighted can be calculated from the class-wise scores; micro (which is accuracy unless labels is provided) and samples (if applicable) can be calculated with another call to prf. WDYT?

apacha commented:
I totally agree with you. It probably would be best to just report all applicable averaging schemes. @jnothman, I gave you write-access to my fork, if you could implement that, since I don't precisely know what you mean with "prf".

jnothman commented:
by prf, I just mean precision_recall_fscore_support. I am capable of
implementing it, but I'd be happy if you gave it a go...

On 12 December 2017 at 20:56, Alexander Pacha <notifications@github.com>
wrote:

> I totally agree with you. It probably would be best to just report all
> applicable averaging schemes. @jnothman <https://github.com/jnothman>, I
> gave you write-access to my fork, if you could implement that, since I
> don't precisely know what you mean with "prf".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10029#issuecomment-351001785>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz61QhzwyNIivAIxijNNyxYn4oa4VHks5s_k2zgaJpZM4QJVDB>
> .
>

apacha committed:
Merge branch 'master-upstream'

apacha committed:
Reporting all averages (micro, macro, weighted and sample (if applicable)) when calling classification_report.
Updated numbers in working_with_text_data tutorial, because the data seems to have changed.

apacha commented:
@jnothman I hope that this feature is ready now. I'm not really sure why travis and appveyor failed, but it seems to be unrelated to my changes.

apacha committed:
Removing code that accidentally slipped into a test.

jnothman reviewed (approved):
LGTM!

jnothman commented:
Please add an entry to the change log at `doc/whats_new/v0.20.rst`. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`

apacha committed:
Adding entry in 'whats new' documentation and crediting myself.

apacha committed:
Dummy-change to trigger rebuild of CI.

jnothman commented:
One more thing: I've been recently frustrated by F1 in my own work. It is not a good metric to optimise for (see e.g. rants by David Powers).

I think it would be good to encourage the use of other measures for binary or multiclass classification. `matthews_corrcoef`, cohen's kappa and perhaps balanced accuracy (not currently implemented for multiclass) are all options. We should report at least one of these if we are going to list metrics in classification report.

jnothman commented:
Never mind that comment. Balanced accuracy is already covered by macro-average recall. Reading up a little more, matthews_corrcoef might be more suited where there isn't one "truth" against a prediction is measured. I suppose I'll have to try influencing use in other ways :)

qinhanmin2014 commented:
We've supported ``output_dict`` for ``classification_report`` (#11160), so I think this PR needs to be updated accordingly. I'm going to mark it as 0.21. @apacha apologies for the inconvenience and thanks for your great work.

jnothman commented:
But it could still get merged for 0.20, given the upcoming sprints, if conflicts were resolved...?

jnothman commented:
I think it remains complementary to #11160

qinhanmin2014 commented:
> But it could still get merged for 0.20, given the upcoming sprints, if conflicts were resolved...?

Of course it can. Solving the conflicts will include supporting dict output for various averages. Change the milestone back to 0.20.

glemaitre commented on doc/whats_new/v0.20.rst:
```diff
@@ -144,6 +144,10 @@ Metrics
 
+- :func:`metrics.classification_report` now reports all applicable averages on
+  the given data, including micro, macro and weighted average as well as samples
+  average for multilabel data. :issue:`4558` by :user:`Alexander Pacha <apacha>`. 
```
The issue number is actually the one of this PR.

glemaitre replied:
The issue number refers to the PR number actually.

glemaitre commented on doc/modules/model_evaluation.rst:
```diff
@@ -584,13 +584,15 @@ and inferred labels::
    >>> target_names = ['class 0', 'class 1', 'class 2']
    >>> print(classification_report(y_true, y_pred, target_names=target_names))
-                precision    recall  f1-score   support
+                 precision    recall  f1-score   support
```
Is the doctest failing before?.

glemaitre replied:
We should probably not change anything if this is not the reason.

jnothman replied:
The existing code adjusts the column width to fit the maximum column. Since we now have a longer "weighted avg" column name, everything needs indentation.

glemaitre commented on doc/tutorial/text_analytics/working_with_text_data.rst:
```diff
@@ -370,19 +370,21 @@ analysis of the results::
-           comp.graphics       0.88      0.97      0.92       389
-                 sci.med       0.94      0.90      0.92       396
-  soc.religion.christian       0.90      0.95      0.93       398
+             alt.atheism       0.97      0.60      0.74       319
```
why those scores changed?

GaelVaroquaux replied:
+1. I don't understand.

jnothman replied:
I agree this is strange. Running the doctests locally I can confirm the former values are the correct ones.

glemaitre commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -731,27 +731,55 @@ def test_classification_report_multiclass():
     # print classification report with class names
     expected_report = """\
-             precision    recall  f1-score   support
+              precision    recall  f1-score   support
```
why the new space?

glemaitre commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -731,27 +731,55 @@ def test_classification_report_multiclass():
+weighted avg       0.33      0.33      0.33         9
+"""
+    report = classification_report(y_true, y_pred)
+    assert_equal(report, expected_report)
```
```python
assert report == expected_report
```

glemaitre commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -731,27 +731,55 @@ def test_classification_report_multiclass():
     report = classification_report(
         y_true, y_pred, labels=np.arange(len(iris.target_names)),
         target_names=iris.target_names)
     assert_equal(report, expected_report)
```
you can update with pytest style

```python
assert report == expected_report
```

glemaitre reviewed (changes requested):
If this is for 0.20, here come some comments.

GaelVaroquaux commented on doc/tutorial/text_analytics/working_with_text_data.rst:
```diff
@@ -370,19 +370,21 @@ analysis of the results::
+  array([[192   2   6 119]
+         [  2 347   4  36]
+         [  2  11 322  61]
+         [  2   2   1 393]])
```
Same thing here: why the change?

GaelVaroquaux reviewed (changes requested):
There are changes that I don't understand. I don't feel comfortable merging this.

amueller committed:
Merge branch 'master' into pr/10029

# Conflicts:
#	sklearn/metrics/classification.py

amueller committed:
fix some doctests in text data

amueller commented:
should be good now?

qinhanmin2014 commented:
I think this PR cannot be merged since it does not support dictionary output. I've opened #11679 to finish the PR.

qinhanmin2014 commented:
Closing in favor of #11679, thanks @apacha
